README.md
# Nebula-Python-SDK

Travis CI unit tests & auto PyPi push status: [![Build Status](https://travis-ci.org/nebula-orchestrator/nebula-python-sdk.svg?branch=master)](https://travis-ci.org/nebula-orchestrator/nebula-python-sdk)

Code coverage: [![codecov](https://codecov.io/gh/nebula-orchestrator/nebula-python-sdk/branch/master/graph/badge.svg)](https://codecov.io/gh/nebula-orchestrator/nebula-python-sdk)

An SDK for managing [nebula](https://nebula-orchestrator.github.io/) via python.

translates all of Nebula [API](https://nebula.readthedocs.io/en/latest/api/general/) calls to something more pythonic.

Source code is available at [github](https://github.com/nebula-orchestrator/nebula-python-sdk)

# How To Use
first get NebulaPythonSDK onto your machine

```bash
# Install from PyPi
pip install NebulaPythonSDK
```

now use it in your code
```python
# Load API wrapper from library
from NebulaPythonSDK import Nebula

# Create API object.
# port defaults to 80, protocol defaults to "http" & request_timeout defaults to 60 if any of them is not set.
connection = Nebula(username="your_nebula_user", password="your_nebula_pass", host="nebula.example.com", port=80, protocol="http", request_timeout=60)

# List apps
app_list = connection.list_apps()

# List app info
app_config = connection.list_app_info("app_name")

# Create app
app_conf = {
    "containers_per_cpu": 8,
    "env_vars": {
        "test": "blabla",
        "test3t2t32": "tesg4ehgee"
    },
    "docker_ulimits": [],
    "networks": ["nebula"],
    "running": True,
    "rolling_restart": False,
    "volumes": ["/tmp:/tmp/1", "/var/tmp/:/var/tmp/1:ro"],
    "containers_per": {
        "cpu": 6
    },
    "starting_ports": [
        {
            "81": 80
        }
    ],
    "docker_image": "httpd",
    "privileged": False,
    "devices": []
}
connection.create_app("app_name", app_conf)

# create device group
device_group_config = {"apps": ["app_name"]}
connection.create_device_group("device_group_name", device_group_config)

# list device group
connection.list_device_group("device_group_name")

# list device group info
connection.list_device_group_info("device_group_name")

# ping api
connection.check_api()

# delete app
connection.delete_app("app_name")

# stop app
connection.stop_app("app_name")

# start app
connection.start_app("app_name")

# restart app
connection.restart_app("app_name")

# update app
connection.update_app("app_name", app_conf)

# update app force all params - will reset all params to default values if not declared and raise an error for params 
# that are required but not given, similar to the POST update of the API
connection.update_app("app_name", app_conf, force_all=True)

# prune images on all device groups
connection.prune_images()

# delete device_group
connection.delete_device_group("device_group_name")

# prune images on a selected device groups
connection.prune__device_group_images("device_group_name")

# update device group
connection.update_device_group("device_group_name", device_group_config)

# list paginated reports
connection.list_reports()

# create a user
user_config = {"password": "user_password", "token": "user_token"}
connection.create_user("user_name", user_config)

# list user info
connection.list_user("user_name")

# list all users
connection.list_users()

# update a user
user_config = {"password": "new_password", "token": "new_token"}
connection.update_user("user_name", user_config)

# refresh a user token
connection.refresh_user_token("user_name")

# delete a user
connection.delete_user("user_name")

# create a user group
user_group_config = {
                        "group_members": [
                            "user_name"
                            ], 
                        "pruning_allowed": True, 
                        "apps": {
                            "app_name": "rw"
                            }, 
                        "device_groups": {
                            "device_group_name": "ro"
                            }, 
                        "admin": False
                    }
connection.create_user_group("user_group_name", user_group_config)

# list user group info
connection.list_user_group("user_group_name")

# list all users group
connection.list_user_groups()

# update a user group
user_group_config = {"admin": False}
connection.update_user_group("user_group_name", user_group_config)

# delete a user group
connection.delete_user_group("user_group_name")

# list all cron jobs
connection.list_cron_jobs()

# delete a cron job
connection.delete_cron_job("cron_job_name")

# list a cron job
connection.list_cron_job_info("cron_job_name")

# create a cron job
cron_job_config = {
  "env_vars": {"test": "test123"},
  "docker_image" : "nginx",
  "running": True,
  "volumes": [],
  "networks": ["nebula", "bridge"],
  "devices": [],
  "privileged": False,
  "schedule": "0 * * * *"
}
connection.create_cron_job("cron_job_name", cron_job_config)

# update a cron job
cron_job_config = {
  "schedule": "5 * * * *"
}
connection.update_cron_job("cron_job_name", cron_job_config)

# update a cron job force all params - will reset all params to default values if not declared and raise an error for 
# params that are required but not given, similar to the POST update of the API
cron_job_config = {
  "schedule": "5 * * * *",
  "docker_image" : "nginx"
}
connection.update_cron_job("cron_job_name", cron_job_config, force_all=True)

```


CHANGELOG.md
# Changelog

## [Unreleased](https://github.com/nebula-orchestrator/nebula-python-sdk/tree/HEAD)

[Full Changelog](https://github.com/nebula-orchestrator/nebula-python-sdk/compare/2.8.0...HEAD)

**Implemented enhancements:**

- Bump docutils from 0.16 to 0.17 [\#137](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/137) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump twine from 3.3.0 to 3.4.1 [\#136](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/136) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump urllib3 from 1.26.3 to 1.26.4 [\#135](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/135) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump pygments from 2.8.0 to 2.8.1 [\#133](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/133) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump tqdm from 4.56.2 to 4.59.0 [\#132](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/132) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump readme-renderer from 28.0 to 29.0 [\#130](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/130) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump pygments from 2.7.4 to 2.8.0 [\#128](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/128) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump tqdm from 4.56.0 to 4.56.2 [\#127](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/127) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump bleach from 3.2.3 to 3.3.0 [\#125](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/125) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump bleach from 3.2.1 to 3.2.3 [\#124](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/124) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump urllib3 from 1.26.2 to 1.26.3 [\#123](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/123) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump pkginfo from 1.6.1 to 1.7.0 [\#121](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/121) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump pygments from 2.7.3 to 2.7.4 [\#120](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/120) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump tqdm from 4.55.1 to 4.56.0 [\#119](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/119) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump tqdm from 4.55.0 to 4.55.1 [\#117](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/117) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump tqdm from 4.54.1 to 4.55.0 [\#115](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/115) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump twine from 3.2.0 to 3.3.0 [\#114](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/114) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump requests from 2.25.0 to 2.25.1 [\#113](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/113) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump chardet from 3.0.4 to 4.0.0 [\#112](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/112) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump tqdm from 4.54.0 to 4.54.1 [\#111](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/111) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump certifi from 2020.11.8 to 2020.12.5 [\#110](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/110) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump pygments from 2.7.2 to 2.7.3 [\#109](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/109) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump tqdm from 4.53.0 to 4.54.0 [\#108](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/108) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump tqdm from 4.52.0 to 4.53.0 [\#107](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/107) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump tqdm from 4.51.0 to 4.52.0 [\#106](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/106) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump urllib3 from 1.26.1 to 1.26.2 [\#105](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/105) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump requests from 2.24.0 to 2.25.0 [\#104](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/104) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump urllib3 from 1.25.11 to 1.26.1 [\#103](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/103) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump certifi from 2020.6.20 to 2020.11.8 [\#101](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/101) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump pkginfo from 1.6.0 to 1.6.1 [\#100](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/100) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump tqdm from 4.50.2 to 4.51.0 [\#99](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/99) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump pygments from 2.7.1 to 2.7.2 [\#98](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/98) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump readme-renderer from 27.0 to 28.0 [\#97](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/97) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump pkginfo from 1.5.0.1 to 1.6.0 [\#96](https://github.com/nebula-orchestrator/nebula-python-sdk/pull/96) ([dependabot[bot]](https://github.com/apps/dependabot))

test/travis_ci_scripts/travis_ci_build_script.sh
#!/usr/bin/env bash

# auth into docker hub to avoid pull rate limiting
echo "$DOCKER_HUB_PASSWORD" | docker login -u "$DOCKER_HUB_USER" --password-stdin
sudo chmod 777 /home/travis/.docker/config.json


# run the following to start the API, MongoDB
sudo docker-compose -f test/travis_ci_scripts/docker-compose.yml pull
sudo docker-compose -f test/travis_ci_scripts/docker-compose.yml up -d

# wait until the manager is online
until $(curl --output /dev/null --silent --head --fail -H 'authorization: Basic bmVidWxhOm5lYnVsYQ==' -H 'cache-control: no-cache' http://127.0.0.1/api/v2/status); do
    echo "Waiting on the manager API to become available..."
    sleep 3
done

echo "manager available"


NebulaPythonSDK/sdk.py
import requests, base64, json, six


def b64encode(source):
    if six.PY3:
        source = source.encode('utf-8')
    content = base64.b64encode(source).decode('utf-8')


class Nebula:

    # the nebula class init module serves as the login against the nebula API as it's the only shared thing among the
    # class functions
    def __init__(self, username=None, password=None, token=None, host="127.0.0.1", port=80, protocol="http",
                 host_uri=None, request_timeout=60):
        self.request_timeout = request_timeout
        self.username = username
        self.password = password
        self.token = token
        self.protocol = protocol
        self.port = port
        if host_uri:
            self.host = host_uri
        else:
            self.host = protocol + "://" + host + ":" + str(port)
        self.headers = {
            'content-type': "application/json",
            'cache-control': "no-cache"
        }
        if token is not None:
            self.headers["authorization"] = "Bearer " + self.token
        elif username is not None and password is not None:
            user_pass_combo = username + ":" + password
            if six.PY3 is True:
                user_pass_combo = user_pass_combo.encode('utf-8')
            self.basic_auth = base64.b64encode(user_pass_combo)
            self.headers["authorization"] = "Basic " + self.basic_auth.decode('utf-8')
        self.API_VERSION = "v2"

    # create a new nebula app, requires the app name to create and a complete dict of the config values for it
    def create_app(self, app, config):
        url = self.host + "/api/" + self.API_VERSION + "/apps/" + app
        payload = json.dumps(config)
        headers = self.headers
        response = requests.request("POST", url, data=payload, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # delete an existing nebula app, no confirmation required in SDK so be careful
    def delete_app(self, app):
        url = self.host + "/api/" + self.API_VERSION + "/apps/" + app
        headers = self.headers
        response = requests.request("DELETE", url, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # list all of the apps managed by nebula
    def list_apps(self):
        url = self.host + "/api/" + self.API_VERSION + "/apps"
        headers = self.headers
        response = requests.request("GET", url, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # list the config of a nebula app, only requires the app name
    def list_app_info(self, app):
        url = self.host + "/api/" + self.API_VERSION + "/apps/" + app
        headers = self.headers
        response = requests.request("GET", url, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # stop a nebula app, only requires the app name
    def stop_app(self, app):
        url = self.host + "/api/" + self.API_VERSION + "/apps/" + app + "/stop"
        headers = self.headers
        response = requests.request("POST", url, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # start a nebula app, only requires the app name
    def start_app(self, app):
        url = self.host + "/api/" + self.API_VERSION + "/apps/" + app + "/start"
        headers = self.headers
        response = requests.request("POST", url, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # restart a nebula app, only requires the app name
    def restart_app(self, app):
        url = self.host + "/api/" + self.API_VERSION + "/apps/" + app + "/restart"
        headers = self.headers
        response = requests.request("POST", url, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # update a nebula app, requires the app name and a dict of the config values you want to change, any combination of
    # config values is accepted as it keeps the rest unchanged
    def update_app(self, app, config, force_all=False):
        if force_all is True:
            request_type = "POST"
        else:
            request_type = "PUT"
        url = self.host + "/api/" + self.API_VERSION + "/apps/" + app + "/update"
        payload = json.dumps(config)
        headers = self.headers
        response = requests.request(request_type, url, data=payload, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # prune unused images on all devices
    def prune_images(self):
        url = self.host + "/api/" + self.API_VERSION + "/prune"
        headers = self.headers
        response = requests.request("POST", url, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # prune unused images on all devices that are part of the device_group given
    def prune__device_group_images(self, device_groups):
        url = self.host + "/api/" + self.API_VERSION + "/device_groups/" + device_groups + "/prune"
        headers = self.headers
        response = requests.request("POST", url, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # check that the contacted api is responding as expected
    def check_api(self):
        url = self.host + "/api/" + self.API_VERSION + "/status"
        headers = self.headers
        response = requests.request("GET", url, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # list the info of all apps and all data of a given device_group
    def list_device_group_info(self, device_group):
        url = self.host + "/api/" + self.API_VERSION + "/device_groups/" + device_group + "/info"
        headers = self.headers
        response = requests.request("GET", url, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # list all device_groups
    def list_device_groups(self):
        url = self.host + "/api/" + self.API_VERSION + "/device_groups"
        headers = self.headers
        response = requests.request("GET", url, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # list device_group configuration
    def list_device_group(self, device_group):
        url = self.host + "/api/" + self.API_VERSION + "/device_groups/" + device_group
        headers = self.headers
        response = requests.request("GET", url, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # create a new nebula device_group, requires the app name to create and a complete dict of the config values for it
    def create_device_group(self, device_group, config):
        url = self.host + "/api/" + self.API_VERSION + "/device_groups/" + device_group
        payload = json.dumps(config)
        headers = self.headers
        response = requests.request("POST", url, data=payload, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # delete an existing nebula device_group, no confirmation required in SDK so be careful
    def delete_device_group(self, device_group):
        url = self.host + "/api/" + self.API_VERSION + "/device_groups/" + device_group
        headers = self.headers
        response = requests.request("DELETE", url, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # update a device group
    def update_device_group(self, device_group, config):
        url = self.host + "/api/" + self.API_VERSION + "/device_groups/" + device_group + "/update"
        payload = json.dumps(config)
        headers = self.headers
        response = requests.request("PUT", url, data=payload, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # list device_group configuration
    def list_reports(self, page_size=10, hostname=None, device_group=None, report_creation_time_filter="gt",
                     report_creation_time=None, last_id=None, updated=None):
        url = self.host + "/api/" + self.API_VERSION + "/reports"
        headers = self.headers
        querystring = {
            "page_size": page_size,
            "hostname": hostname,
            "device_group": device_group,
            "report_creation_time_filter": report_creation_time_filter,
            "report_creation_time": report_creation_time,
            "last_id": last_id,
            "updated": updated
        }
        response = requests.request("GET", url, headers=headers, timeout=self.request_timeout, params=querystring)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # list all users
    def list_users(self):
        url = self.host + "/api/" + self.API_VERSION + "/users"
        headers = self.headers
        response = requests.request("GET", url, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # list a user configuration
    def list_user(self, user):
        url = self.host + "/api/" + self.API_VERSION + "/users/" + user
        headers = self.headers
        response = requests.request("GET", url, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # delete an existing nebula user, no confirmation required in SDK so be careful
    def delete_user(self, user):
        url = self.host + "/api/" + self.API_VERSION + "/users/" + user
        headers = self.headers
        response = requests.request("DELETE", url, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # update a user pass and\or bearer token
    def update_user(self, user, config):
        url = self.host + "/api/" + self.API_VERSION + "/users/" + user + "/update"
        payload = json.dumps(config)
        headers = self.headers
        response = requests.request("PUT", url, data=payload, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # refresh a token for a user
    def refresh_user_token(self, user):
        url = self.host + "/api/" + self.API_VERSION + "/users/" + user + "/refresh"
        headers = self.headers
        response = requests.request("POST", url, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # create a new nebula user, requires the user name to create and a complete dict of the config values for it
    def create_user(self, user, config):
        url = self.host + "/api/" + self.API_VERSION + "/users/" + user
        payload = json.dumps(config)
        headers = self.headers
        response = requests.request("POST", url, data=payload, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # list all user groups
    def list_user_groups(self):
        url = self.host + "/api/" + self.API_VERSION + "/user_groups"
        headers = self.headers
        response = requests.request("GET", url, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # list a user group configuration
    def list_user_group(self, user):
        url = self.host + "/api/" + self.API_VERSION + "/user_groups/" + user
        headers = self.headers
        response = requests.request("GET", url, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # delete an existing nebula user group, no confirmation required in SDK so be careful
    def delete_user_group(self, user):
        url = self.host + "/api/" + self.API_VERSION + "/user_groups/" + user
        headers = self.headers
        response = requests.request("DELETE", url, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # update a user group
    def update_user_group(self, user, config):
        url = self.host + "/api/" + self.API_VERSION + "/user_groups/" + user + "/update"
        payload = json.dumps(config)
        headers = self.headers
        response = requests.request("PUT", url, data=payload, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # create a new nebula user group, requires the group name to create and a complete dict of the config values for it
    def create_user_group(self, user, config):
        url = self.host + "/api/" + self.API_VERSION + "/user_groups/" + user
        payload = json.dumps(config)
        headers = self.headers
        response = requests.request("POST", url, data=payload, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # list all of the cron_jobs managed by nebula
    def list_cron_jobs(self):
        url = self.host + "/api/" + self.API_VERSION + "/cron_jobs"
        headers = self.headers
        response = requests.request("GET", url, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # delete an existing nebula cron_job, no confirmation required in SDK so be careful
    def delete_cron_job(self, cron_job):
        url = self.host + "/api/" + self.API_VERSION + "/cron_jobs/" + cron_job
        headers = self.headers
        response = requests.request("DELETE", url, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # list the config of a nebula cron_job, only requires the app name
    def list_cron_job_info(self, cron_job):
        url = self.host + "/api/" + self.API_VERSION + "/cron_jobs/" + cron_job
        headers = self.headers
        response = requests.request("GET", url, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # create a new nebula cron_job, requires the cron_job name to create and a complete dict of the config values for it
    def create_cron_job(self, cron_job, config):
        url = self.host + "/api/" + self.API_VERSION + "/cron_jobs/" + cron_job
        payload = json.dumps(config)
        headers = self.headers
        response = requests.request("POST", url, data=payload, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response

    # update a nebula cron_job, requires the cron_job name and a dict of the config values you want to change, any
    # combination of config values is accepted as it keeps the rest unchanged
    def update_cron_job(self, cron_job, config, force_all=False):
        if force_all is True:
            request_type = "POST"
        else:
            request_type = "PUT"
        url = self.host + "/api/" + self.API_VERSION + "/cron_jobs/" + cron_job + "/update"
        payload = json.dumps(config)
        headers = self.headers
        response = requests.request(request_type, url, data=payload, headers=headers, timeout=self.request_timeout)
        filtered_response = {"status_code": response.status_code, "reply": response.json()}
        return filtered_response
